driver-kotlin-sync/src/main/kotlin/com/mongodb/kotlin/client/MongoClient.kt

```diff
@@ -0,0 +1,282 @@
+     * @throws IllegalArgumentException if databaseName is invalid
+     * @see com.mongodb.MongoNamespace.checkDatabaseNameValidity
+     */
+    public fun getDatabase(databaseName: String): MongoDatabase = MongoDatabase(wrapped.getDatabase(databaseName))
+
+    /**
+     * Creates a client session.
+     *
+     * Note: A ClientSession instance can not be used concurrently in multiple operations.
+     *
+     * @param options the options for the client session
+     * @return the client session
+     */
+    public fun startSession(options: ClientSessionOptions = ClientSessionOptions.builder().build()): ClientSession =
+        ClientSession(wrapped.startSession(options))
+
+    /**
+     * Get a list of the database names
+     *
+     * @return an iterable containing all the names of all the databases
+     * @see [List Databases](https://www.mongodb.com/docs/manual/reference/command/listDatabases)
+     */
+    public fun listDatabaseNames(): MongoIterable<String> = MongoIterable(wrapped.listDatabaseNames())
+
+    /**
+     * Gets the list of databases
+     *
+     * @param clientSession the client session with which to associate this operation
+     * @return the list databases iterable interface
+     * @see [List Databases](https://www.mongodb.com/docs/manual/reference/command/listDatabases)
+     */
+    public fun listDatabaseNames(clientSession: ClientSession): MongoIterable<String> =
+        MongoIterable(wrapped.listDatabaseNames(clientSession.wrapped))
+
+    /**
+     * Gets the list of databases
+     *
+     * @return the list databases iterable interface
+     */
+    @JvmName("listDatabasesAsDocument")
+    public fun listDatabases(): ListDatabasesIterable<Document> = listDatabases<Document>()
+
+    /**
+     * Gets the list of databases
+     *
+     * @param clientSession the client session with which to associate this operation
+     * @return the list databases iterable interface
+     */
+    @JvmName("listDatabasesAsDocumentWithSession")
+    public fun listDatabases(clientSession: ClientSession): ListDatabasesIterable<Document> =
+        listDatabases<Document>(clientSession)
+
+    /**
+     * Gets the list of databases
+     *
+     * @param T the type of the class to use
+     * @param resultClass the target document type of the iterable.
+     * @return the list databases iterable interface
+     */
+    public fun <T : Any> listDatabases(resultClass: Class<T>): ListDatabasesIterable<T> =
+        ListDatabasesIterable(wrapped.listDatabases(resultClass))
+
+    /**
+     * Gets the list of databases
+     *
+     * @param T the type of the class to use
+     * @param clientSession the client session with which to associate this operation
+     * @param resultClass the target document type of the iterable.
+     * @return the list databases iterable interface
+     */
+    public fun <T : Any> listDatabases(clientSession: ClientSession, resultClass: Class<T>): ListDatabasesIterable<T> =
```

Comment:
you can use the inline reified trick here (with @JvmName)

Reply:
Ack - will try to work around it. The main issue is inlined methods cannot be called from Java code.  However, we should still be able test the resulting MongoIterables just we'd have to construct them manually.

Reply:
@jyemin - turns out removing the resultClass methods has a number of consequences.

- The wrapped MongoClient / MongoDatabase / MongoCollections then have to become publicly visible. (There is no package private in Kotlin) and reified functions require public access. (https://kotlinlang.org/docs/inline-functions.html#restrictions-for-public-api-inline-functions)
- The syncadapter cannot call the reified methods - so essentially you have to manually construct the MongoIterables by calling the wrapped methods directly. So doesn't actually  test the full API.
- The API becomes hard to wrap by 3rd parties. Reified types in methods are inlined at compile time, so like the sync adapter users with a Class<T> cannot access the reified method. Having an overload is the way around that.

Only suggestion I have it we could half the overloaded API size by making clientSession optional for all the methods that take a resultClass. eg: 

```
public fun <T : Any> listDatabases(clientSession: ClientSession?, resultClass: Class<T>): ListDatabasesIterable<T>
```

Reply:
Note, you can use `@PublishedApi` to allow reified methods calling internal methods: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-published-api/

Reply:
Can we add a resolution to this thread, please?

Reply:
The initial comment is resolved (using inline trick)

As @cmelchior mentions you can use `@PublishedApi` annotation when using reified methods and internal values. Which can be used for my first point (As long as the syncadapter is moved into the same package namespace).

However, the removal of resultClass has wider implications as mentioned in my thread - so should remain - so that users can call them if they ever need to at runtime.